src/librustc/traits/select.rs

```diff
@@ -2630,7 +2630,13 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
                                            snapshot)
         });
 
-        obligations.extend(trait_obligations);
+        for trait_obligation in trait_obligations.into_iter() {
+            // Since auto traits are coinductive we can skip having a predicate with a nested
+            // obligation equal to itself.
+            if trait_obligation.predicate != ty::Predicate::Trait(obligation.predicate) {
```

Comment:
Is this `if` statement here for a reason?

Reply:
`trait_obligations` will almost always contain `obligation` (which comes from `impl_or_trait_obligations` in line 2624), creating a cycle.  This cycle will eventually be ignored by the obligation forest anyway, so this is not necessary to fix issue #40827.

So, this is part optimization (don't generate obligations that are trivially satisfied (given that we are talking about coinductive predicates)), and part a fix in the behavior of the function (it doesn't make much sense to say that `A: Send` is a sub-obligation of `A: Send`, even if that would be trivially acceptable).

Reply:
@orium are you available on some chat?